### ISSUE TITLE
Menu block on content type "Guide" does not drive content in the middle of the page

### ISSUE THREAD

**Issue by mashanadavis** [NIHL/Drupal7-Migration#13]
This was a requirement. 
The steps taken were:
1. created Guide
2. created associated webpages
3. created menu block and assigned to Guide

Here's an example of the functionality: http://www.nibib.nih.gov/about-nibib/strategic-plan

**Comment by boharaganga** [comment]
http://nihlibrarystg.prod.acquia-sites.com/node/326

**Comment by jronmorgan** [comment]
I don't understand. What is the actual problem?

**Comment by gbohara** [comment]
http://nihlibrarystg.prod.acquia-sites.com/node/326
If you go to this page, there are Links on the left side
ie. 
  About the Technology Lab
  Equipment, Software , & Tools
  ....
  ....

Clicking on these links should only change the content on the right side of this menu. This menu should still be visible.

**Comment by gbohara** [comment]
The nodes that are linked from the links are not of type "guide".
For example: 
The link "Data Visualization" on the left side links to "/node/621" which is not a "guide".

If a "guide" is created by setting proper Node-level-blocks  and then the block is configured to link to that "guide" node, it should work.

Mashana, Please let me know if you need more instructions. I would be happy to help.

**Comment by mashanadavis** [comment]
I will need help doing this.

**Comment by mashanadavis** [comment]
Ganga - So, I don't create pages using the Webpage content type and add these pages to a block and then assign the block to the Guide? I believe that's what I did.

**Comment by gbohara** [comment]
What are these links? How often they change? Who is responsible for editing these contents? 
One simple way to solve this problem is to create Guides for all the content that needs to be in those links. (If it does not change quite often and only content admin are responsible for managing these content, it makes more sense this way)

Further research is needed if we want to be able to link anything (guide + other pages) in these side links.

**Comment by mashanadavis** [comment]
The links are relative to the Guide. For example, Technology Hub services guide may have links for data visualization, 3D printing, and so on. We also want to link to relevant training or a list of books, journals, and databases. The last set of information would be controlled by a taxonomy (NIH Subjects) and linked to using a View. 

Creating a guide for each page may work, but you will have to make sure that the same in-page menu block and other content---whatever is in the right and left sidebars---is duplicated for each Guide.  The way that it should work is that the links to the  webpages in the in-page menu block should live in the guide (has the same URL path). Is there a way to add webpages within a Guide? Could these fields be added to this content type, and then served up as a block on the Guide?

I go back to the example I linked in the initial ticket: http://www.nibib.nih.gov/about-nibib/strategic-plan. This is a Drupal site and it has the same functionality we are looking to achieve. Not sure if you can see the code. We know the developer for this site. I can ask how it was done. 

FYI, our content authors (mostly non-technical people) will manage these pages. So, the functionality and creation of this content has to be simple.

**Comment by tgifford-webfirst** [comment]
The NIBIB pages are using a standard Drupal Main Menu (capitalized to denote the Drupal entity). It appears all content is listed only once in the Menu. That single Menu is then broken down with the menu_blocks module into the top/main menu and the sidebar menu, which shows the 2nd level of menu items. This is *much* easier to use than the current nihl approach, which uses the node-level blocks module to place blocks on specific pages. However, it does not work if you need to have content referenced from multiple places within the Menu. In other words, if content can be referenced from more than one Guide.

To determine the best approach, we should develop a logical sitemap that lists the various menus and where content should be placed within them.

**Comment by mashanadavis** [comment]
Whatever approach is simpler for our content authors. We assumed that the functionality would be the same as we referenced this site when spelling out the requirements. 

I will create a sitemap for these pages if that will help to determine the approach.

**Comment by tgifford-webfirst** [comment]
This is related to #10 and #57. I think having a sitemap will help on all these tickets. Let's track that in a basecamp discussion.

**Comment by svpatel1** [comment]
after speaking with the team, we feel it is important to have the full sitemap before we can proceed with this ticket. thank you!

**Comment by mashanadavis** [comment]
I uploaded a site map to Basecamp: https://basecamp.com/2765815/projects/7275781/messages/49592632. Ted and I are discussing this.

**Comment by tgifford-webfirst** [comment]
This can be implemented with the menu_block module, which provides functionality for selecting the menu to show based on which menus contain the current page, sorted by an administrator-defined order. By default, this menu order is static (always the same), but we can customize it based on the current user's agency in a hook_init implementation by overriding the order in the $conf array. Combined with menu_block's advanced configuration, we can render from the e.g. 2nd-level of the menu or the level of the current page as appropriate by site section.

**Comment by cmndr-king** [comment]
Bridget agreed to pull together the list of menu items and target information for Tech Hub and Bioinformatics so that Ted has enough information to flesh out two service pages.

**Comment by tgifford-webfirst** [comment]
Document has been added to basecamp. Needs review by WebFirst.

**Comment by tgifford-webfirst** [comment]
Secondary menu has been added in all appropriate contexts, and is driven by the links in the main menu. Addition of this menu introduces the need to review the "nodeblock" menus previously added to pages such as http://nihlibrarystg.prod.acquia-sites.com/node/326. I will create a new ticket to review those pages, and list them there.

**Comment by tgifford-webfirst** [comment]
Mashana will add links at a lower level to check the functionality.

**Comment by mashanadavis** [comment]
Ted - Adding the lower level options for Bioinformatics in the global menu and on the actual service page is fine. I like this for the other menu options as well. And, I see where this can be done in TB Mega Menus. We probably need a little CSS to clean up the display on both as it is difficult to distinguish between the higher and lower level options on the services page.

The real issue lies with what's displayed when the lower level option is clicked. Our goal was to have the content of the page change without taking the user away from the page (like NIBIB). When I click on the link to Bioinformatics Consults, it takes me to the web form out of the services landing page. The same action occurs for the Tech Hub options. 

Is it because I used the content type Webpage for the subpages? 

I cannot remember if this is in another ticket.

**Comment by svpatel1** [comment]
http://nihlibrarystg.prod.acquia-sites.com/node/591

**Comment by tgifford-webfirst** [comment]
There are a couple things going on here, which I will describe in detail below.

* Multiple links to the same node (or path) in the same menu
* The secondary menu is not presently shown for Webforms
* The Tech Hub options are a special case

# Multiple links to the same node (or path) in the same menu

When a path, such as for the Bioinformatics Consult request form, is linked multiple times in a Drupal menu, Drupal must choose one menu link to use to generate the breadcrumb and the secondary menu. Unfortunately, Drupal doesn't make this choice explicit--the menu link that is chosen depends on the database implementation, in our case MySql, and other technical details. In short, it is likely that the last link you added to the menu will be the one chosen. If you wanted to switch which link was chosen, you could simply swap the location of the latest link, or delete and re-create the link in the location you want Drupal to choose.

There is still a chance that Drupal/MySql will pick the "wrong" menu link, in which case we would need to find a workaround, such as:
* Create a redirect to the desired path, then put the redirect as the path for the "wrong" link.
* Write custom code to override the menu_link_get_preferred static cache entry for the path in question
* Write custom code to modify the query to MySql, see https://www.drupal.org/node/1649062

This is why we previously recommended only linking to content once in the main menu.

# The secondary menu is not presently shown for Webforms

When I first implemented the secondary menu, all the webforms were listed under Services > Requests, and they weren't linked elsewhere in the menu, so the requests menu was shown as the secondary menu whenever one of the linked webforms was viewed. That didn't seem useful to me because users are not likely to make several different types of requests at once. In addition, the original context settings did not include a secondary menu (i.e. from the prior developer). For those reasons, I did not show the secondary menu for Webforms.

If we make a policy of adding a second link to all Webforms within their relevant menu structures (as you did with Bioinformatics Consult), then we could show the secondary menu, and it would remain consistent.

# The Tech Hub links are a special case

The Tech Hub links are not part of the secondary menu, they are from a menu in a Node-level block (a block that is selected to show on that specific node via the node edit form). In addition, most of the links in that Tech Hub menu seem to be broken.

Based on the sitemap, the proposed links within Tech Hub, such as Data Visualization, are also going to be linked elsewhere (e.g. Data Services Program). This takes us back to the first issue regarding multiple links with the same path.

It is also possible to have an "alternate" navigation menu or menus to show on certain pages. This might be used for cross-cutting topics, such as Data Visualization or other related services. This menu (or menus) might be called "Related Topics", "More Like This", etc. Of course, such a menu might be very helpful to users, but is more effort for the site administrator to maintain.

**Comment by tgifford-webfirst** [comment]
Need styles for lowest levels, then can mark for QA.
